src/main/java/mate/academy/service/impl/AuthenticationServiceImpl.java

```diff
@@ -0,0 +1,72 @@
+package mate.academy.service.impl;
+
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.util.Optional;
+import mate.academy.exception.AuthenticationException;
+import mate.academy.exception.RegistrationException;
+import mate.academy.lib.Inject;
+import mate.academy.lib.Service;
+import mate.academy.model.User;
+import mate.academy.service.AuthenticationService;
+import mate.academy.service.UserService;
+
+@Service
+public class AuthenticationServiceImpl implements AuthenticationService {
+
+    private static final String HASHING_ALGORITHM = "SHA-256";
+    @Inject
+    private UserService userService;
+
+    public AuthenticationServiceImpl(UserService userService) {
+        this.userService = userService;
+    }
+
+    @Override
+    public User login(String email, String password) throws AuthenticationException {
+        Optional<User> optionalUser = userService.findByEmail(email);
+        if (optionalUser.isEmpty()) {
+            throw new AuthenticationException("User not found by email: " + email);
+        }
+        User user = optionalUser.get();
+        String hashedPassword = hashPassword(password, user.getSalt());
+        if (!user.getPassword().equals(hashedPassword)) {
+            throw new AuthenticationException("Invalid password");
+        }
+        return user;
+    }
+
+    @Override
+    public User register(String email, String password) throws RegistrationException {
+        if (userService.findByEmail(email).isPresent()) {
+            throw new RegistrationException("Email is already taken: " + email);
+        }
+        byte[] salt = generateSalt();
+        String hashedPassword = hashPassword(password, salt);
+        User user = new User(email, hashedPassword);
+        user.setSalt(salt);
+        return userService.add(user);
+    }
+
+    private String hashPassword(String password, byte[] salt) {
+        try {
+            MessageDigest digest = MessageDigest.getInstance(HASHING_ALGORITHM);
+            digest.update(salt);
+            byte[] hashedBytes = digest.digest(password.getBytes());
+            StringBuilder hashedPassword = new StringBuilder();
+            for (byte b : hashedBytes) {
+                hashedPassword.append(String.format("%02x", b));
+            }
+            return hashedPassword.toString();
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException("Could not hash password", e);
```

Comment:
Consider using a more specific exception than `RuntimeException` for hashing errors. This can help in better error handling and debugging.